### ISSUE TITLE
[BUG] FFT based model produce divergent training result with Adam

### ISSUE THREAD

**Issue by ChenkaiMao97** [ml-explore/mlx#2366]
**Describe the bug**
Hi, I want to use FFT based model on Apple Silicon, and I've encountered divergent training results using MLX or the mps backend in torch (it's convergent on torch with Cuda or cpu). I have an issue opened for torch as well for reference: https://github.com/pytorch/pytorch/issues/151740

Basically, I observe that when I use FFT based model together with Adam (and also AdamW, Adamax, RMSprop, Lion etc.), I observe the loss goes down, then explode. If I switch to a convolution based model, or if I use SGD (or Adagrad or AdaDelta), the result becomes convergent (but I do want to use Adam for better performance).

The learning rate I used is 1e-4, which should be very small, For smaller lr like 1e-5, the blow up happens after ~25000 steps. On MLX, the good thing is CPU and GPU seem to have good agreement. 

I tried Adam with and without bias_correction, and the blowup happens for both.

One observation I have is, if I set the eps in adam to be very high, for example 0.1, then it seems blowup does not happen. I'm wondering if the normalization in adam, i.e. numerator / (mx.sqrt(v) + eps) will be numerically unstable, especially with FFT. I'm not sure how to dig deeper and debug. 
 
**To Reproduce**
Include code snippet
```python
import mlx.nn as nn
import mlx.core as mx
import numpy as np
import mlx.optimizers as optim

def MSE_loss(a,b):
    return mx.mean((a-b)**2)

################ model definition ##################
class FFTModel(nn.Module):
    def __init__(self, in_channels, out_channels, modes1, modes2, seed=0):
        super().__init__()
        self.in_channels = in_channels
        self.out_channels = out_channels
        self.modes1 = modes1
        self.modes2 = modes2

        scale = (1 / in_channels / out_channels)
        self.weights = mx.array(mx.random.uniform(
            low=0, high=scale,
            shape=(in_channels, out_channels, modes1, modes2, 2),
            key=mx.random.key(seed)
        ))

    def compl_mul2d(self, input_ft, weight_ft):
        # input_ft: (B, Cin, H, W), weight_ft: (Cin, Cout, H, W)
        return mx.einsum('bihw,iohw->bohw', input_ft, weight_ft)

    def __call__(self, x):
        _, _, height, width = x.shape
        
        # design choice 1: use weights as low frequencies
        full_weights = mx.zeros((self.in_channels, self.out_channels, height, width//2 + 1, 2))
        full_weights[:,:,:self.modes1, :self.modes2,:] = self.weights
        complex_w = full_weights[..., 0] + 1j*full_weights[..., 1]
        
        # rfft
        x_ft = mx.fft.rfft2(x, axes=(-2, -1))  # (B, C, H, W//2+1)
        # Complex multiplication
        out_ft = self.compl_mul2d(x_ft, complex_w)

        # iFFT2
        out = mx.fft.irfft2(out_ft, s=(height, width), axes=(-2, -1))  # back to spatial
        return out

class ConvModel(nn.Module):
    def __init__(self, in_c, out_c):
        super().__init__()
        self.conv = nn.Conv2d(in_c, out_c, 3,padding=1)
    
    def __call__(self, x):
        x = x.transpose(0, 2, 3, 1)
        x = self.conv(x)
        x = x.transpose(0, 3, 1, 2)
        return x

#################### training with different backends ################
batch_size = 16
in_c = 2
out_c = 2
lr = 1e-4
sizex, sizey = (128, 128)
modes1, modes2 = (32, 32)

def train(backend, seed = 42):
    np.random.seed(seed)
    if backend=='cpu':
        mx.set_default_device(mx.cpu)
    elif backend=='gpu':
        mx.set_default_device(mx.gpu)
    
    # two model choices
    model = FFTModel(in_c, out_c, modes1, modes2)
    # model = ConvModel(in_c, out_c)


    mx.eval(model.parameters())

    optimizer = optim.Adam(learning_rate=lr, bias_correction=True)
    # optimizer = optim.SGD(learning_rate=lr) # if use SGD, results converge but performance is worse

    criterion = MSE_loss

    x_train = mx.array(mx.random.normal((batch_size, in_c, sizex, sizey), key=mx.random.key(seed)))
    y_train = mx.array(mx.random.normal((batch_size, out_c, sizex, sizey), key=mx.random.key(seed)))

    def loss_fn(model, x_train, y_train):
        out = model(x_train) 
        loss = criterion(out, y_train)
        return loss

    for step in range(5000):
        loss_and_grad_fn = nn.value_and_grad(model, loss_fn)
        loss, grads = loss_and_grad_fn(model, x_train, y_train)
        optimizer.update(model, grads)

        if (step+1) % 500 == 0:
            print(f"Step {(step+1):03d} | Loss: {loss.item():.6f}")

print("training on gpu:")
train('gpu')

print("training on cpu:")
train('cpu')
```

**Expected behavior**
This is my output:

training on gpu:
Step 500 | Loss: 0.969161
Step 1000 | Loss: 0.965178
Step 1500 | Loss: 0.964372
Step 2000 | Loss: 0.966398
Step 2500 | Loss: 0.970965
Step 3000 | Loss: 0.977841
Step 3500 | Loss: 0.986852
Step 4000 | Loss: 0.997869
Step 4500 | Loss: 1.010795
Step 5000 | Loss: 1.025560
training on cpu:
Step 500 | Loss: 0.969155
Step 1000 | Loss: 0.965170
Step 1500 | Loss: 0.964365
Step 2000 | Loss: 0.966391
Step 2500 | Loss: 0.970955
Step 3000 | Loss: 0.977835
Step 3500 | Loss: 0.986846
Step 4000 | Loss: 0.997861
Step 4500 | Loss: 1.010790
Step 5000 | Loss: 1.025555

**Desktop (please complete the following information):**
 - OS Version: Sequoia 15.2
 - Version MLX: 0.26.3, python: 3.10.17

**Additional context**
I'm running on a mac studio (M2 Ultra).

**Comment by angeloskath** [comment]
Thanks for finding this. I pushed a fix at #2367. However, looking at the code above it is quite inefficient to do the 2x2 matmul with einsum. You can write it simply as 

```python
def mul2d_sum(a, b):
    return (a[:, :, None] * b).sum(1)
```

or

```python
@mx.compile
def _mul2d_inner(x1, x2, w11, w12, w21, w22):
    return x1 * w11 + x2 * w21, x1 * w12 + x2 * w22

def mul2d_elw(a, b)
    a, b = _inner_mul2d(x[:, 0], x[:, 1], w[0, 0], w[0, 1], w[1, 0], w[1, 1])
    return mx.stack([a, b], axis=1)
```

or the fastest one would be via a small custom kernel. I haven't benchmarked the above but I would go for the 1st one given it is the simplest possible way to write this.

**Comment by ChenkaiMao97** [comment]
Thanks for the quick response! I have another finding is, if I modify the complex weight to be real only:
complex_w = full_weights[..., 0] + 1j*0
then it looks like the gradient is correct. 

If I modify it to be 
complex_w =0 + 1j*full_weights[..., 1]
then it will blowup, and looks like if I flip the gradient sign, it becomes convergent. 

Basically the gradient for the imaginary part seems wrong.

It looks like you've already pointed this to imag() vjp, which I think could be the problem.

**Comment by angeloskath** [comment]
Yes exactly the gradient for `imag()`  is wrong and it is being used implicitly in the matmul.
